peers/app_request_network.go

```diff
@@ -103,83 +101,105 @@ func NewNetwork(
-	// gossip will enable connecting to all the remaining nodes in the network.
-	var beaconIPs, beaconIDs []string
+	validatorClient := validators.NewCanonicalValidatorClient(logger, pChainClient)
 
-	peers, err := infoClient.Peers(context.Background())
-	if err != nil {
-		logger.Error(
-			"Failed to get peers",
-			zap.Error(err),
-		)
-		return nil, nil, err
+	arNetwork := &AppRequestNetwork{
+		Network:         testNetwork,
+		Handler:         handler,
+		infoClient:      infoClient,
+		logger:          logger,
+		lock:            new(sync.Mutex),
+		validatorClient: validatorClient,
 	}
 
-	// Randomly select peers to connect to until we have numInitialTestPeers
-	indices := rand.Perm(len(peers))
-	for _, index := range indices {
-		// Do not attempt to connect to private peers
-		if len(peers[index].PublicIP) == 0 {
-			continue
-		}
-		beaconIPs = append(beaconIPs, peers[index].PublicIP)
-		beaconIDs = append(beaconIDs, peers[index].ID.String())
-		if len(beaconIDs) == numInitialTestPeers {
-			break
+	// Manually connect to the validators of each of the source subnets.
+	// We return an error if we are unable to connect to sufficient stake on any of the subnets.
+	// Sufficient stake is determined by the Warp quora of the configured supported destinations,
+	// or if the subnet supports all destinations, by the quora of all configured destinations.
+	for _, sourceBlockchain := range cfg.SourceBlockchains {
+		// Get the list of destination blockchains that this source may relay to.
+		// If supported destinations are not specified, use the configured destinations.
+		var destinationBlockchainIDs []ids.ID
+		if supportedDsts := sourceBlockchain.GetSupportedDestinations(); supportedDsts.Len() > 0 {
+			destinationBlockchainIDs = supportedDsts.List()
+		} else {
+			for _, dst := range cfg.DestinationBlockchains {
+				destinationBlockchainIDs = append(destinationBlockchainIDs, dst.GetBlockchainID())
+			}
 		}
-	}
-	if len(beaconIPs) == 0 {
-		logger.Error(
-			"Failed to find any peers to connect to",
-			zap.Error(err),
-		)
-		return nil, nil, err
-	}
-	if len(beaconIPs) < numInitialTestPeers {
-		logger.Warn(
-			"Failed to find a full set of peers to connect to on startup",
-			zap.Int("connectedPeers", len(beaconIPs)),
-			zap.Int("expectedConnectedPeers", numInitialTestPeers),
-		)
-	}
 
-	for i, beaconIDStr := range beaconIDs {
-		beaconID, err := ids.NodeIDFromString(beaconIDStr)
-		if err != nil {
-			logger.Error(
-				"Failed to parse beaconID",
-				zap.String("beaconID", beaconIDStr),
-				zap.Error(err),
-			)
-			return nil, nil, err
+		var connectedValidators *ConnectedCanonicalValidators
+		// If the source blockchain is the primary network, connect to the validators of
+		// the destination blockchains in the following loop.
+		// Otherwise, connect to the validators of the source blockchain
+		if sourceBlockchain.GetSubnetID() != constants.PrimaryNetworkID {
```

Comment:
Could we simplify this logic by moving the case for the primary network and the case for non-primary networks out to helper functions?

For non-primary sources, we:
- connect to source
- iterate over each destination to check the quorum value

For primary sources we:
- iterate over each destination and create connections
- check the connection reaches a quorum for that destination

Reply:
We could also add in a `checkSufficientStakeForDestination` helper used by both of those that takes a `*ConnectedCanonicalValidators` and a single destination chain ID.

Reply:
Sure thing. Added these helpers.